pkg/fqdn/service/service.go

```diff
@@ -0,0 +1,164 @@
+	}
+
+	go func() {
+		<-s.ctx.Done()
+		s.log.Info("FQDN service context done, cleaning up resources")
+		if s.ctx.Err() != nil {
+			s.log.Error("FQDN service context error: ", logfields.Error, s.ctx.Err())
+		}
+		s.closeServer()
+	}()
+
+	return s
+}
+
+// OnIPIdentityCacheChange is a method to receive the IP identity cache change events
+func (s *FQDNDataServer) OnIPIdentityCacheChange(modType ipcache.CacheModification, cidr types.PrefixCluster, oldHostIP, newHostIP net.IP, oldID *ipcache.Identity, newID ipcache.Identity, encryptKey uint8, k8sMeta *ipcache.K8sMetadata, endpointFlags uint8) {
+	s.identityToIpMutex.Lock()
+	switch modType {
+	case ipcache.Upsert:
+		ip := cidr.AsIPNet().IP
+		s.currentIdentityToIp[newID.ID] = append(s.currentIdentityToIp[newID.ID], ip)
+	case ipcache.Delete:
+		if oldID != nil {
+			delete(s.currentIdentityToIp, oldID.ID)
+		}
+	}
+	s.identityToIpMutex.Unlock()
+}
+
+// UpdatePolicyRulesLocked updates the current state of the DNS rules with the given policies and sends the current state of the DNS rules to the client
+// This method is called:
+// 1. when the DNS rules are updated during the endpoint regeneration, we store the state of the DNS rules with flag rulesUpdate as true
+// 2. when the client subscribes to DNS policies, we send the current state of the DNS rules to the client(flag rulesUpdate as false)
+// 3. when the IP identity cache changes, we update the current state of the identity to IP mapping and send the current state of the DNS rules to
+// the client(flag rulesUpdate as false)
+func (s *FQDNDataServer) UpdatePolicyRulesLocked(policies map[identity.NumericIdentity]policy.SelectorPolicy, rulesUpdate bool) error {
+	s.snapshotMutex.Lock()
+	defer s.snapshotMutex.Unlock()
+
+	return nil
+}
+
+// UpdateMappingRequest updates the FQDN mapping with the given data
+// SDP sends the fqdn mapping to cilium agent
+// Steps to update the mapping:
+// 1. Get the endpoint from the IP
+// 2. If the endpoint is not found, return an error
+// 3. If the IPs are not empty, update the cilium agent with the mapping
+func (s *FQDNDataServer) UpdateMappingRequest(ctx context.Context, mappings *pb.FQDNMapping) (*pb.UpdateMappingResponse, error) {
+	return &pb.UpdateMappingResponse{
+		Response: pb.ResponseCode_RESPONSE_CODE_NO_ERROR,
+	}, nil
+}
+
+// RunServer starts the Standalone DNS Proxy grpc server on the given port
+func RunServer(port int, server *FQDNDataServer) error {
+	address := fmt.Sprintf("localhost:%d", port)
+	server.log.Info("Starting Standalone DNS Proxy server on: ", logfields.Address, address)
+	lis, err := net.Listen("tcp", address)
+	if err != nil {
+		server.log.Error("Failed to listen: ", logfields.Error, err)
+		return err
+	}
+	grpcServer := grpc.NewServer(grpc.KeepaliveEnforcementPolicy(kaep), grpc.KeepaliveParams(kasp))
+	pb.RegisterFQDNDataServer(grpcServer, server)
+
+	closer := func() {
+		grpcServer.GracefulStop()
+	}
+	server.closeServer = closer
```

Comment:
This write here is a data race. You are reading it from a go routine spawned by `NewServer` without any synchronization between the read (in `NewServer`) and the write here. You need to add synchronization between to two: https://go.dev/ref/mem